http-client-core/src/main/java/io/micronaut/http/client/HttpVersionSelection.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.http.HttpVersion;
+import io.micronaut.http.client.annotation.Client;
+
+import java.util.Arrays;
+
+/**
+ * This class collects information about HTTP client protocol version settings, such as the
+ * {@link PlaintextMode} and the ALPN configuration.
+ *
+ * @author Jonas Konrad
+ * @since 4.0
+ */
+public final class HttpVersionSelection {
```

Comment:
perhaps this can be a record since we are in the Java 17 world

Reply:
i don't want to expose the constructor. e.g. if alpn is false, alpnSupportedProtocols makes no sense, so the constructor is a bit confusing api.